assertj-error-prone/src/main/java/com/palantir/assertj/errorprone/AssertjAssertThatThrownBy.java

```diff
@@ -0,0 +1,118 @@
+            Matchers.expressionStatement(
+                    MethodMatchers.staticMethod().onClass("org.junit.Assert").named("fail")),
+            Matchers.expressionStatement(MethodMatchers.staticMethod()
+                    .onClass("org.junit.jupiter.api.Assertions")
+                    .named("fail")),
+            Matchers.expressionStatement(MethodMatchers.staticMethod()
+                    .onClass("org.assertj.core.api.Assertions")
+                    .named("fail")));
+
+    @Override
+    public Description matchTry(TryTree tree, VisitorState state) {
+        List<? extends StatementTree> tryStatements = tree.getBlock().getStatements();
+        if (tryStatements.isEmpty() || tree.getCatches().size() != 1 || tree.getFinallyBlock() != null) {
+            return Description.NO_MATCH;
+        }
+        CatchTree catchTree = Iterables.getOnlyElement(tree.getCatches());
+        if (!catchTree.getBlock().getStatements().isEmpty()
+                || catchTree.getParameter().getType().getKind() == Tree.Kind.UNION_TYPE) {
+            return Description.NO_MATCH;
+        }
+        StatementTree lastStatement = Iterables.getLast(tryStatements);
+        List<? extends StatementTree> throwingStatements = tryStatements.subList(0, tryStatements.size() - 1);
+        if (!FAIL_METHOD.matches(lastStatement, state) || throwingStatements.isEmpty()) {
+            return Description.NO_MATCH;
+        }
+        if (!TestCheckUtils.isTestCode(state)) {
+            return Description.NO_MATCH;
+        }
+        Optional<String> failMessage = getFailMessage(lastStatement, state);
+        return buildDescription(tree)
+                .addFix(tryFailToAssertThatThrownBy(
+                        tree, throwingStatements, catchTree.getParameter(), failMessage, state))
+                .build();
+    }
+
+    private static Optional<String> getFailMessage(StatementTree failStatement, VisitorState state) {
+        Iterable<? extends ExpressionTree> failArgs =
+                ((MethodInvocationTree) ((ExpressionStatementTree) failStatement).getExpression()).getArguments();
+        return Optional.ofNullable(Iterables.get(failArgs, 0, null))
+                .map(state::getSourceForNode)
+                .filter(msg -> !IGNORED_FAIL_MESSAGES.contains(msg));
+    }
+
+    private static Optional<Fix> tryFailToAssertThatThrownBy(
+            TryTree tree,
+            List<? extends StatementTree> throwingStatements,
+            VariableTree catchParameter,
+            Optional<String> failMessage,
+            VisitorState state) {
+        SuggestedFix.Builder fix = SuggestedFix.builder();
+        fix.addStaticImport("org.assertj.core.api.Assertions.assertThatThrownBy");
+        fix.replace(
+                ((JCTree) tree).getStartPosition(),
+                ((JCTree) throwingStatements.iterator().next()).getStartPosition(),
+                "assertThatThrownBy(() -> {");
```

Comment:
Do you think we could special case blocks reduced to a single line and leave out the curly braces?

Reply:
This is already done by the formatter which runs after all the fixes. I would prefer not adding this logic in code and leave it up for the formatter.

Reply:
I don't think our formatter knows how to convert between statement and expression lambdas because there's potential for ambiguity.

Reply:
You are right! I added the check if the try-block contains only a single statement (not counting the `fail()`). In this case, the single-line lambda is now used.